### PR TITLE
[tune] TensorFlow Distributed Trainable

### PR DESCRIPTION
**Pull request by oliverhu**
## Why are these changes needed?
Follow on the discussion on https://github.com/ray-project/ray/discussions/11111. Tune currently supports distributed hyperparameter search with distributed training for (pytorch, horovod). We should also support Tensorflow default distributed training strategies including MultiWorkerMirrorStrategy and Parameter Serving Strategy.

## Related issue number
https://github.com/ray-project/ray/discussions/11111

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by oliverhu** [comment]
Sample run output: https://gist.github.com/oliverhu/dca9f4d722e0f516ac8aba9701c59eee

File: python/ray/tune/integration/tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    """Converts TensorFlow MultiWorkerMirror training to be executable by Tune.
+
+        Requires TensorFlow > 2.0 to work, recommends TensorFlow > 2.2.
+
+        This function wraps and sets resources for a TF distributed training
+        function to be used with Tune. It generates a TensorFlow Trainable
+        which can be a distributed
+        training job.
+
+        Note: there is no fault tolerance at the moment.
```
Comment (richardliaw) [inline python/ray/tune/integration/tensorflow.py:119]: Nit: only 4 space indent.

File: python/ray/tune/examples/tf_distributed_keras_example.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                                                              0.5)),
+        metrics=["accuracy"])
+    return model
+
+
+def train_mnist(config, checkpoint_dir=None):
+    strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
+    per_worker_batch_size = 64
+    tf_config = json.loads(os.environ["TF_CONFIG"])
+    num_workers = len(tf_config["cluster"]["worker"])
```
Comment (richardliaw) [inline python/ray/tune/examples/tf_distributed_keras_example.py:49]: hmm, can we avoid this? maybe have a:

```
ray.tune.integration.tensorflow.get_num_workers()
```
?

Comment (oliverhu) [reply python/ray/tune/examples/tf_distributed_keras_example.py:49]: Updated.. this is more like a common util function for all distributed tf training tho.

Comment (richardliaw) [reply python/ray/tune/examples/tf_distributed_keras_example.py]: yeah, i guess you could do integration.tensorflow.util.get_num_workers

**Comment by richardliaw** [comment]
Thanks a bunch for this PR! 

Can you be sure to add the following:
 * Run the test in the test suite similar to this (https://github.com/ray-project/ray/pull/9739/files#diff-5a9919a777b978fee388f8d0477558ecb309ade3bdb68f0cb400a13be873cddeR290)
 * Add documentation (like this https://github.com/ray-project/ray/pull/9739/files#diff-e28682a144e210b167be5d875472c84ee9a69add6036a6ccf966394949115d75R288)
 * Add the example to the docs like (https://github.com/ray-project/ray/pull/9739/files#diff-c6b9448a9f2448ee8cb89d555bd600092ee8305b6e4d267dc9290d48201fe3ccR44)

**Comment by oliverhu** [comment]
@richardliaw thanks for detailed comments.. updated

**Review by richardliaw — APPROVED** [review]

**Pull request merged**